skorch/classifier.py

```diff
@@ -103,7 +103,8 @@ def check_data(self, X, y):
     # pylint: disable=arguments-differ
     def get_loss(self, y_pred, y_true, *args, **kwargs):
         if isinstance(self.criterion_, torch.nn.NLLLoss):
-            y_pred = torch.log(y_pred)
+            eps = torch.finfo(y_pred.dtype).eps
+            y_pred = torch.log(y_pred + eps)
```

Comment:
This is **not** a smart operation.
Running `log(exp( · ))` is not computationally stable.

Reply:
Since this is guarded by the `isinstance(self.criterion_, torch.nn.NLLLoss)`, this only happens with this negative log loss. You can use a different `criterion_` such as `CrossEntropyLoss`, and skorch would not take this log.

Fundamentally, this comes from a constraint where the output of `module(X)` are probabilities: https://github.com/skorch-dev/skorch/issues/481#issuecomment-497454767

> The idea was that at that time, it was standard (keras, nolearn) that if you define a classifier, your output nonlinearity should be softmax.

Reply:
@Atcold did you have any practical issues resulting from this?

Reply:
I was following these tutorials ([1](https://skorch.readthedocs.io/en/stable/user/quickstart.html#training-a-model) and [2](https://github.com/skorch-dev/skorch/blob/master/notebooks/MNIST.ipynb)), which I've assigned to my mentee, but then was surprised to see the soft(arg)max inside the model. So, I've looked up the source, and saw that nasty thing 😬 
I have not run any code, but it's bad practice. Especially using it in an example / tutorial.

**If** you really like a soft(arg)max inside a model (since this appears to be *standard*, for some), then I'd rather get the logit, by indexing the `L - 1` layer, and use that in a `CrossEntropy` criterion.

Reply:
We discussed the issue at length (see the linked comment above). There are other design decisions that we need to take into account (how to go from `module_.forward` to `net.predict_proba` for instance), so you cannot judge this decision in isolation. And if we changed it now, it would be a backwards incompatible change, breaking a lot of existing code. Moreover, I have yet to encounter a problem with that -- if this really causes numerical instability, it almost certainly means that you've done something wrong with your model in the first place. Finally, if you change the loss, you'll not run into the problem, so it's easy to circumvent if it should really cause trouble. For all those reasons, my point of view is to leave it as it is.

Reply:
Okay, added my 2¢ there.